controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/AddHostTask.java

```diff
@@ -0,0 +1,229 @@
+                     int hostPort,
+                     String hostAZ,
+                     String hdfsDir,
+                     int rateLimitMbs) {
+    this(new Param()
+        .setHostToAdd(
+            new HostBean()
+                .setIp(hostIp)
+                .setPort(hostPort)
+                .setAvailabilityZone(hostAZ)
+        )
+        .setHdfsDir(hdfsDir)
+        .setRateLimitMbs(rateLimitMbs)
+    );
+  }
+
+  public AddHostTask(Param param) {
+    super(param);
+  }
+
+  @Override
+  public void process(Context ctx) throws Exception {
+    final String clusterName = ctx.getCluster();
+    final String hdfsDir = getParameter().getHdfsDir();
+    final HostBean hostToAdd = getParameter().getHostToAdd();
+    final int rateLimitMbs = getParameter().getRateLimitMbs();
+
+    InetSocketAddress hostAddr = new InetSocketAddress(hostToAdd.getIp(), hostToAdd.getPort());
+    Admin.Client client = clientFactory.getClient(hostAddr);
+
+    // 1) ping the host to add to make sure it's up and running.
+    try {
+      client.ping();
+      // continue if #ping() succeeds.
+    } catch (TException tex) {
+      ctx.getTaskQueue().failTask(ctx.getId(), "Host to add is not alive!");
+      return;
+    }
+
+    ClusterBean clusterBean = ZKUtil.getClusterConfig(zkClient, clusterName);
+    if (clusterBean == null) {
+      ctx.getTaskQueue().failTask(ctx.getId(), "Failed to read cluster config from zookeeper.");
+      return;
+    }
+
+    for (SegmentBean segment : clusterBean.getSegments()) {
+      // 2) find shards to serve for new host
+      Set<Integer> shardToServe =
+          IntStream.range(0, segment.getNumShards())
+                   .boxed()
+                   .collect(Collectors.toSet());
+      for (HostBean host : segment.getHosts()) {
+        // ignore hosts in different AZ than the new host
+        if (host.getAvailabilityZone().equals(hostToAdd.getAvailabilityZone())) {
+          host.getShards().forEach(shard -> shardToServe.remove(shard.getId()));
+        }
+      }
+
+      // 3) upload shard data to the new host
+      for (int  shardId : shardToServe) {
+        HostBean upstream = findMasterShard(shardId, segment.getHosts());
+        if (upstream == null) {
+          //TODO: should we fail the task in this case?
+          LOG.error("Failed to find master shard for segment={}, shardId={}",
+                    segment.getName(), shardId);
+          continue;
+        }
+        Admin.Client upstreamClient = clientFactory.getClient(
+            new InetSocketAddress(upstream.getIp(), upstream.getPort())
+        );
+        String dbName = ShardUtil.getDBNameFromSegmentAndShardId(segment.getName(), shardId);
+        String hdfsPath = String.format("%s/%s/%s/%05d/%s/%s", hdfsDir, clusterName,
+            segment.getName(), shardId, upstream.getIp(), getCurrentDateTime());
+
+        upstreamClient.backupDB(new BackupDBRequest(dbName, hdfsPath).setLimit_mbs(rateLimitMbs));
```

Comment:
do we want to catch exception? for example process is not running at this moment